File: python/utils/get_all_domains.py
from typing import List

from solana.rpc.async_api import AsyncClient
from solana.rpc.types import MemcmpOpts, DataSliceOpts
from solders.pubkey import Pubkey

from constants import ROOT_DOMAIN_ACCOUNT, NAME_PROGRAM_ID


async def get_all_domains(
    connection: AsyncClient,
    wallet: Pubkey,
) -> List[Pubkey]:
    filters = [
        MemcmpOpts(
            offset=32,
            bytes=str(wallet),
        ),
        MemcmpOpts(
            offset=0,
            bytes=str(ROOT_DOMAIN_ACCOUNT),
        ),
    ]
    data_slice = DataSliceOpts(offset=0, length=0)

    account = await connection.get_program_accounts(
        pubkey=NAME_PROGRAM_ID,
        filters=filters,
        data_slice=data_slice,
    )
    return [elem.pubkey for elem in account.value]

File: js/src/utils/getAllDomains.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAllDomains = void 0;
const constants_1 = require("../constants");
/**
 * This function can be used to retrieve all domain names owned by `wallet`
 * @param connection The Solana RPC connection object
 * @param wallet The wallet you want to search domain names for
 * @returns
 */
function getAllDomains(connection, wallet) {
    return __awaiter(this, void 0, void 0, function* () {
        const filters = [
            {
                memcmp: {
                    offset: 32,
                    bytes: wallet.toBase58(),
                },
            },
            {
                memcmp: {
                    offset: 0,
                    bytes: constants_1.ROOT_DOMAIN_ACCOUNT.toBase58(),
                },
            },
        ];
        const accounts = yield connection.getProgramAccounts(constants_1.NAME_PROGRAM_ID, {
            filters,
            // Only the public keys matter, not the data
            dataSlice: { offset: 0, length: 0 },
        });
        return accounts.map((a) => a.pubkey);
    });
}
exports.getAllDomains = getAllDomains;
